test/unit/DominantJuice.unit.t.sol

```diff
@@ -0,0 +1,962 @@
+            address(directory),
+            abi.encodeCall(directory.isTerminalOf, (projectID, ethPaymentTerminal)),
+            abi.encode(true)
+        );
+
+        // Pledger1 tries to redeem again via JB.
+        vm.prank(address(ethPaymentTerminal));
+        vm.expectRevert(DominantJuice.AlreadyWithdrawnRefund.selector);
+        dominantJuice.didRedeem(didRedeemData);
+    }
+
+    // PASSING
+    function testFuzz_didRedeem_happyPath(uint256 _pledgeAmount) public bonusDeposited {
+        // Pledger1 pledges up to and excluding cycleTarget amount and time advances past expiry date.
+        _pledgeAmount = bound(_pledgeAmount, MIN_PLEDGE_AMOUNT, CYCLE_TARGET - 1);
+        pledge(pledger1, _pledgeAmount);
+        vm.warp(cycleExpiryDate + 100);
+
+        didRedeemData.holder = pledger1;
+        didRedeemData.projectId = projectID;
+
+        vm.mockCall(
+            address(directory),
+            abi.encodeCall(directory.isTerminalOf, (projectID, ethPaymentTerminal)),
+            abi.encode(true)
+        );
+        vm.expectCall(address(directory), abi.encodeCall(directory.isTerminalOf, (projectID, ethPaymentTerminal)));
+
+        (uint256 pledgerWeight, uint256 totalWeight) = dominantJuice.getPledgerAndTotalWeights(pledger1);
+        DominantJuice.Campaign memory campaign = dominantJuice.getCampaignInfo();
+
+        uint256 pledgerRefundBonus = (pledgerWeight / totalWeight) * campaign.totalRefundBonus;
+
+        if (dominantJuice.getBalance() < pledgerRefundBonus) {
+            vm.expectRevert(abi.encodeWithSelector(DominantJuice.InsufficientFunds.selector));
+        }
+
+        vm.expectEmit(true, true, true, true, address(dominantJuice));
+        emit CycleRefundBonusWithdrawal(pledger1, pledgerRefundBonus);
+
+        vm.prank(address(ethPaymentTerminal));
+        dominantJuice.didRedeem(didRedeemData);
+
+        assertEq(true, dominantJuice.getPledgerRefundStatus(pledger1));
+        assertEq(0, dominantJuice.getBalance());
+        assertEq(pledgerRefundBonus, TOTAL_REFUND_BONUS);
+        // Since pledger1 pays through the JB architecture, it doesn't subtract from pledger1's balance.
+        // The unit test shouldn't cover outside state, but this was just an additional sanity assert.
+        assertEq(STARTING_BALANCE + pledgerRefundBonus, pledger1.balance);
+    }
+
+    // Add more didRedeem fuzz tests with multiple redeemers
+
+    // EDGE CASE - contract has less balance than it should. How to mock?
+
+    //////////////////////////////
+    // creatorWithdraw() Tests
+    //////////////////////////////
+
+    function test_creatorWithdraw_revertsForNonOwner() public bonusDeposited {
+        // vm.assume(_notOwner != campaignManager && _notOwner != address(0));
+        successfulCycleHasExpired();
+
+        vm.expectRevert(
+            "AccessControl: account 0x8e24d86be44ab9006bd1277bddc948ecebbfbf6c is missing role 0x5022544358ee0bece556b72ae8983c7f24341bd5b9483ce8a19bff5efbb2de92"
```

Comment:
is there a way to check part of the string or a regular expression? I typically just check `missing role`

Reply:
Not sure. I’ll do research on that and report back.

Reply:
Couldn't figure this one out, so changed these tests to explicit, non-fuzzed.